core/shared/src/main/scala/spire/math/interval/Overlap.scala

```diff
@@ -0,0 +1,90 @@
+package spire.math.interval
+
+import spire.algebra.{Eq, Order}
+import spire.math.Interval
+
+/**
+  * An ADT that represents overlapping result for any two intervals.
+  * All subclasses and method names are "left-biased":
+  * for example, [[StrictlyLess]] result means, that [[lhs]] is strictly less than [[rhs]].
+  */
+sealed abstract class Overlap[A: Order] extends Serializable {
+  def lhs: Interval[A]
+  def rhs: Interval[A]
+  def isStrictlyLess: Boolean
```

Comment:
I think these could all be defined up-front here, with some pattern matching. I wonder if we also need all 4 - seems like the important ones `isDisjoint`, `isSubset`, and `isEqual`. If it is LessAndOverlaps or StrictlyLess, then both `isEqual` and `isSubset` will be false, and `isDisjoint` indicates the particular case.